README.md

```diff
@@ -216,34 +215,109 @@ If you register the service providers in the correct order in your container (A
+
+Consider the code below:
+
+```php
+class MyServiceProvider implements ServiceProvider
+{
+    public function getServices()
+    {
+        return [
+            'my_service' => function() {
+                return new MyService();
+            }
+        ];
+    }
+}
+```
+
+When asked the `my_service` service, a cached/compiled container will have to instantiate the service provider, then call the `getServices` method.
 
-#### Why static methods?
+In PHP, the minimal way to fetch the service is:
 
-It comes from the assumption that container configuration is stateless, i.e. it doesn't depend on anything.
+```php
+$serviceProvider = new MyServiceProvider();
+$factories = $serviceProvider->getServices();
+$service = $factories['my_service']();
+```
 
-If container configuration is not stateless, it introduces problems for compiled containers (e.g. Symfony) or containers that perform caching (e.g. PHP-DI). Indeed, compiled containers and cached containers will call the `getServices` method in a process and the factories in another process. Relying on some state of the service provider class might therefore be a bad idea.
+Now, let's have a look at this similar code:
 
-Using static methods *conveys* the idea that everything should be stateless. Furthermore, since factories are supposed to be stateless, there is no need for a factory to rely on the service provider state (using `$this`), hence the idea to make factories static.
+```php
+class MyServiceProvider implements ServiceProvider
+{
+    public function getServices()
+    {
+        return [
+            'my_service' => [ MyServiceProvider::class, 'createMyService' ],
+        ];
+    }
+    
+    public static function createMyService()
+    {
+        return new MyService();
+    }
+}
+```
 
-Also static methods are easy to call, no need to instantiate any object which, when building a container in every request, is a bit of a performance advantage compared to having to create new objects for no reason.
+Compiled/cached container can now run `getServices()` once and subsequently directly refer to the factory method (`MyServiceProvider::createMyService`).
 
-#### Why does the `getServices()` method returns an array of function names instead of an array of callables?
+Now, the minimal code to fetch the service becomes:
 
-An array of callables means anonymous functions could be returned, which is hard to adapt to compiled or cached containers.
+```php
+$service = MyServiceProvider::createMyService()();
+```
 
-Even if we explicitly forbid closures (which would be weird in the first place), invokable objects or object methods would also hard to adapt with compiled/cached containers.
+So the second solution offers more room for optimization for compiled/cached containers.
 
-So the only callables that would really be allowed would be:
+As a courtesy for users looking for the best performance, service providers may consider using `public static function`s as factories.
```

Comment:
I think we could skip that too. Either we encourage it straight into the standard or we don't mention it *yet*. It will also conflict with definition objects.